File: CustomLabel.vb
Imports System.Drawing
Imports System.Windows.Forms

Public Class CustomLabel
    Inherits Label

    Private _borderColor As Color = Color.Black
    Private _borderWidth As Double = 0.1

    Public Property BorderColor As Color
        Get
            Return _borderColor
        End Get
        Set(value As Color)
            _borderColor = value
            Invalidate() ' Redraw the control
        End Set
    End Property

    Public Property BorderWidth As Integer
        Get
            Return _borderWidth
        End Get
        Set(value As Integer)
            _borderWidth = value
            Invalidate() ' Redraw the control
        End Set
    End Property

    Protected Overrides Sub OnPaint(e As PaintEventArgs)
        MyBase.OnPaint(e)

        Dim borderRect As Rectangle = New Rectangle(
            New Point(_borderWidth, _borderWidth),
            New Size(ClientSize.Width - (2 * _borderWidth), ClientSize.Height - (2 * _borderWidth))
        )

        ' Draw border
        Using borderPen As New Pen(_borderColor, _borderWidth)
            e.Graphics.DrawRectangle(borderPen, borderRect)
        End Using
    End Sub
End Class

File: CustomLabel.cs
using System;
using System.Drawing;
using System.Windows.Forms;

namespace VSTO_Addins
{

    public class CustomLabel : Label
    {

        private Color _borderColor = Color.Black;
        private double _borderWidth = 0.1d;

        public Color BorderColor
        {
            get
            {
                return _borderColor;
            }
            set
            {
                _borderColor = value;
                Invalidate(); // Redraw the control
            }
        }

        public int BorderWidth
        {
            get
            {
                return (int)Math.Round(_borderWidth);
            }
            set
            {
                _borderWidth = value;
                Invalidate(); // Redraw the control
            }
        }

        protected override void OnPaint(PaintEventArgs e)
        {
            base.OnPaint(e);

            var borderRect = new Rectangle(new Point((int)Math.Round(_borderWidth), (int)Math.Round(_borderWidth)), new Size((int)Math.Round(ClientSize.Width - 2d * _borderWidth), (int)Math.Round(ClientSize.Height - 2d * _borderWidth)));

            // Draw border
            using (var borderPen = new Pen(_borderColor, (float)_borderWidth))
            {
                e.Graphics.DrawRectangle(borderPen, borderRect);
            }
        }
    }
}